video/image_writer.c

```diff
@@ -262,11 +309,188 @@ static bool write_jpeg(struct image_writer_ctx *ctx, mp_image_t *image, FILE *fp
+    avctx->pkt_timebase = (AVRational){1, 30};
+    avctx->codec_type = AVMEDIA_TYPE_VIDEO;
+    avctx->pix_fmt = imgfmt2pixfmt(image->imgfmt);
+
+    av_opt_set_int(avctx, "still-image", 1, AV_OPT_SEARCH_CHILDREN);
+
+    AVDictionary *avd = NULL;
+    mp_set_avdict(&avd, ctx->opts->avif_opts);
+    av_opt_set_dict2(avctx, &avd, AV_OPT_SEARCH_CHILDREN);
+    av_dict_free(&avd);
+
+    pic = av_frame_alloc();
+    HANDLE_ERROR(!pic, "Could not allocate AVFrame\n");
+
+    for (int n = 0; n < 4; n++) {
+        pic->data[n] = image->planes[n];
+        pic->linesize[n] = image->stride[n];
+    }
+    pic->format = avctx->pix_fmt;
+    pic->width = avctx->width;
+    pic->height = avctx->height;
+    avctx->flags |= AV_CODEC_FLAG_GLOBAL_HEADER; // this...
+    avctx->color_range = pic->color_range =
+        mp_csp_levels_to_avcol_range(image->params.color.levels);
+    avctx->color_primaries = pic->color_primaries =
+        mp_csp_prim_to_avcol_pri(image->params.color.primaries);
+    avctx->color_trc = pic->color_trc =
+        mp_csp_trc_to_avcol_trc(image->params.color.gamma);
+    avctx->colorspace = pic->colorspace =
+        mp_csp_to_avcol_spc(image->params.color.space);
+    avctx->chroma_sample_location = pic->chroma_location =
+        AVCHROMA_LOC_UNSPECIFIED;
+
+    ret = avcodec_open2(avctx, codec, NULL);
+    HANDLE_ERROR(ret < 0, "Could not open libavcodec encoder for saving images\n");
+
+    ret = avio_open(&avioctx, filename, AVIO_FLAG_WRITE);
+    HANDLE_ERROR(ret < 0, "Could not open file '%s' for saving images\n", filename);
+
+    fmtctx = avformat_alloc_context();
+    HANDLE_ERROR(!fmtctx, "Could not allocate format context\n");
+    fmtctx->pb = avioctx;
+    fmtctx->oformat = ofmt;
+
+    stream = avformat_new_stream(fmtctx, codec);
+    HANDLE_ERROR(!stream, "Could not allocate stream\n");
+
+    ret = avcodec_parameters_from_context(stream->codecpar, avctx);
+    HANDLE_ERROR(ret < 0, "Could not copy parameters from context\n");
+
+    MP_DBG(ctx, "write_avif() Codec Context:\n"
+        "  color_trc = %s\n"
+        "  color_primaries = %s\n"
+        "  color_range = %s\n"
+        "  colorspace = %s\n"
+        "  chroma_sample_location = %s\n",
+        av_color_transfer_name(avctx->color_trc),
+        av_color_primaries_name(avctx->color_primaries),
+        av_color_range_name(avctx->color_range),
+        av_color_space_name(avctx->colorspace),
+        av_chroma_location_name(avctx->chroma_sample_location)
+    );
+
+    ret = avformat_init_output(fmtctx, NULL);
+    HANDLE_ERROR(ret < 0, "Could not initialize output\n");
+
+    ret = avformat_write_header(fmtctx, NULL);
+    HANDLE_ERROR(ret < 0, "Could not write format header\n");
+
+    pkt = av_packet_alloc();
+    HANDLE_ERROR(!pkt, "Could not allocate packet\n");
+
+    ret = avcodec_send_frame(avctx, pic);
+    HANDLE_ERROR(ret < 0, "Error sending frame\n");
+
+    int pts = 1;
+
+    static char dbgbuff[129];
+#define DEBUG_SIDE_DATA(x, n) do { \
```

Comment:
Shouldn't you make this a function instead? This is kind of weird.

Reply:
I agree, I'll hoist it out to a function.

Reply:
done.